s3/s3.cc

```diff
@@ -0,0 +1,625 @@
+                // UploadPart: https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+                auto part_number_ = part_number;
+                http_response resp = co_await make_http_request(con, "PUT",
+                    format("{}?partNumber={}&uploadId={}", _path, part_number, _upload_id), _client_impl._headers, std::move(wr));
+                _parts[part_number_].etag = resp.get_header("ETag");
+            }).finally([op = _uploads_barrier.start(), units = std::move(units)] {});
+        }
+
+        future<> start_upload() {
+            _upload_id = co_await _client_impl.do_with_connection([this] (connection_ptr con) -> future<sstring> {
+                // CreateMultipartUpload: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html
+                http_response resp = co_await make_http_request(con, "POST", _path + "?uploads", _client_impl._headers);
+                co_return parse_xml_tag(resp.content_as_string(), "UploadId");
+            });
+        }
+
+        // Aborts the upload.
+        // Static so that the can be invoked from the destructor and doesn't depend on the instance to be alive.
+        // No flushes can be in progress after this in called.
+        static future<> do_abort_upload(client_ptr c, client_impl& impl, sstring path, sstring upload_id) {
+            // AbortMultipartUpload: https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html
+            co_await impl.do_with_connection([&impl, path, upload_id] (connection_ptr con) mutable -> future<> {
+                co_await make_http_request(con, "DELETE", format("{}?uploadId={}", path, upload_id), impl._headers);
+            });
+            log.debug("Upload ({}) of {} aborted.", upload_id, path);
+            // keep c alive until here
+        }
+
+        // Aborts the upload.
+        // Precondition: !upload_active()
+        future<> abort_upload() {
+            log.debug("Aborting upload {} of {}", _upload_id, _path);
+            co_await do_abort_upload(_client, _client_impl, _path, _upload_id);
+            _upload_done = true;
+        }
+
+        // Completes the upload.
+        // Precondition: upload_active()
+        // No flushes can be in progress when this in called.
+        future<> complete_upload_on_server() {
+            // CompleteMultipartUpload: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html
+            memory_data_sink_buffers bufs;
+            {
+                output_stream<char> out(data_sink(std::make_unique<memory_data_sink>(bufs)));
+                co_await out.write("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n");
+                co_await out.write("<CompleteMultipartUpload xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">");
+                uint64_t id = 0;
+                for (part& p : _parts) {
+                    co_await out.write("<Part>");
+                    co_await out.write("<ETag>");
```

Comment:
Nitpick: Why call out.write() for every little string constant instead of calling it once for several, e.g., "<Part><ETag>"?